pkg/apis/core/types.go

```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
 	// +kubebuilder:validation:MaxLength=316
 	Error *string
 }
+
+// PodResourceClaim references exactly one ResourceClaim, either by name or
+// by embedding a template for a ResourceClaim that will get created
+// by the resource claim controller in kube-controller-manager.
+type PodResourceClaim struct {
+	// A name under which this resource can be referenced by the containers.
+	Name string
+
+	// Claim determines where to find the claim.
+	Claim ClaimSource
+}
+
+// ClaimSource either references one separate ResourceClaim by name or
+// embeds a template for a ResourceClaim, but never both.
+//
+// Additional options might get added in the future, so code using this
+// struct must error out when none of the options that it supports are set.
+type ClaimSource struct {
+	// The resource is independent of the Pod and defined by
+	// a separate ResourceClaim in the same namespace as
+	// the Pod. Either this or Template must be set, but not both.
+	ResourceClaimName *string
+
+	// Will be used to create a stand-alone ResourceClaim to allocate the resource.
+	// The pod in which this PodResource is embedded will be the
+	// owner of the ResourceClaim, i.e. the ResourceClaim will be deleted together with the
+	// pod.  The name of the ResourceClaim will be `<pod name>-<resource name>` where
+	// `<resource name>` is the name PodResource.Name
+	// Pod validation will reject the pod if the concatenated name
+	// is not valid for a ResourceClaim (for example, too long).
+	//
+	// An existing ResourceClaim with that name that is not owned by the pod
+	// will *not* be used for the pod to avoid using an unrelated
+	// resource by mistake. Scheduling is then blocked until
+	// the unrelated ResourceClaim is removed. If such a pre-created ResourceClaim is
+	// meant to be used by the pod, the ResourceClaim has to be updated with an
```

Comment:
I don't think we should make it mutable. The user should create a new pod.

Reply:
We cannot prevent the user from manually creating a ResourceClaim that looks like it was created automatically. The API server cannot prevent it because to it a user and the controller look alike (both have permission to create ResourceClaims).

But I guess we don't need to discuss this corner case here. The important part is that the owner reference must be there. I've removed the "If such a pre-created ..." part.